File: benchmarks/pb_pythoninstances/mono.php
<?php

class Terminator {
	function is_terminator() { return True; }
}

class Chain {
	function __construct($value, $next) {
		$this->value = $value;
		$this->next = $next;
	}

	function is_terminator() { return False; }
	function get_value() { return $this->value; }
	function get_next() { return $this->next; }
}

function make_instchain($x) {
	$cur = new Terminator();
	while($x >= 0) {
		$cur = new Chain($x, $cur);
		$x -= 1;
	}

	return $cur;
}

function consume_instchain($chain) {
	$res = 0;
	while(! ($chain->is_terminator())) {
		$res += $chain->get_value();
		$chain = $chain->get_next();
	}
	return $res;
}

function outer($outer, $inner) {
	$correct = floor($inner * ($inner + 1) / 2);
	for ($i = 0; $i < $outer; $i++) {
		$res = consume_instchain(make_instchain($inner));
		assert($res == $correct);
	}
}

function run_iter($n) {
	outer($n, 10000);
}

run_iter(1000);
?>

File: benchmarks/pb_instchain/mono.py
class Terminator(object):

 def is_terminator(self):
    return True

class Chain(object):

    def __init__(self, value, next):
        self.value = value
        self.next = next

    def is_terminator(self):
        return False

    def get_value(self):
        return self.value

    def get_next(self):
        return self.next

def make_instchain(x):
    cur = Terminator()
    while x >= 0:
        cur = Chain(x, cur)
        x -= 1
    return cur

def consume_instchain(chain):
    res = 0
    while not chain.is_terminator():
        res += chain.get_value()
        chain = chain.get_next()
    return res

def outer(outer, inner):
    correct = inner * (inner + 1) // 2
    i = 0
    while i < outer:
        res = consume_instchain(make_instchain(inner))
        assert(res == correct)
        i += 1

def run_iter(n):
    outer(n, 10000)
